plugin/rename_file.py

```diff
@@ -0,0 +1,156 @@
+        return True
+
+    def want_event(self) -> bool:
+        return False
+
+    def input(self, args: dict) -> sublime_plugin.TextInputHandler | None:
+        if "new_name" in args:
+            return None
+        view = self.window.active_view()
+        old_path = args.get('old_path') or view.file_name() if view else None
+        return RenamePathInputHandler(Path(old_path or "").name)
+
+    def run(self, new_name: str, old_path: str | None = None) -> None:
+        view = self.window.active_view()
+        old_path = old_path or view.file_name() if view else None
+        if old_path is None:  # handle renaming buffers
+            if view:
+                view.set_name(new_name)
+            return
+        if new_name == old_path:
+            return
+        # new_name can be: FILE_NAME.xy OR ./FILE_NAME.xy OR ../../FILE_NAME.xy
+        resolved_new_path = (Path(old_path).parent / new_name).resolve()
+        new_path = str(resolved_new_path)
+        if resolved_new_path.exists() and not self.is_case_change(old_path, new_path):
+            self.window.status_message('Unable to Rename. Already exists')
+            return
+        sublime.set_timeout_async(lambda: self.run_async(old_path, new_path))
+
+    def run_async(self, old_path: str, new_path: str) -> None:
+        file_rename: FileRename = {
+            "newUri": filename_to_uri(new_path),
+            "oldUri": filename_to_uri(old_path)
+        }
+        promises = list(self.create_will_rename_requests_async(file_rename))
+        if promises:
+            Promise.all(promises).then(lambda responses: self.handle_rename_async(responses, file_rename))
+        else:
+            self.rename_path(file_rename)
+
+    def create_will_rename_requests_async(
+        self, file_rename: FileRename
+    ) -> Generator[Promise[tuple[WorkspaceEdit | None, weakref.ref[Session]]]]:
+        for session in self.sessions():
+            filters = session.get_capability('workspace.fileOperations.willRename.filters') or []
+            if match_file_operation_filters(filters, file_rename['oldUri']):
+                yield session.send_request_task(Request.willRenameFiles({'files': [file_rename]})) \
+                    .then(partial(lambda weak_session, response: (response, weak_session), weakref.ref(session)))
+
+    def is_case_change(self, path_a: str, path_b: str) -> bool:
+        return path_a.lower() == path_b.lower() and Path(path_a).stat().st_ino == Path(path_b).stat().st_ino
+
+    def handle_rename_async(self, responses: list[tuple[WorkspaceEdit | None, weakref.ref[Session]]],
+                            file_rename: FileRename) -> None:
+        promises: list[Promise[Any]] = []
+        for response, weak_session in responses:
+            if (session := weak_session()) and response:
+                promises.append(session.apply_workspace_edit_async(response, is_refactoring=True)
+                                .then(lambda _: self.rename_path(file_rename)))
```

Comment:
Actually we should also defer `self.rename_path` until all edits are applied. Otherwise in a hypothetical case when two servers provide edits, we'll rename the file already on handling first response and then handling the second response will fail/crash.